test/parse.jl

```diff
@@ -76,3 +77,265 @@ end
+        end
+
+        # Leap year
+        for (j, d) in zip([1; 59:61; 365], [1; 59; 61:62; 366])
+            f, i = TimeZones._parsesub_tzdate("J$j")
+            @test f(2020) == Date(2020) + Day(d - 1)
+            @test dayofyear(f(2020)) == d
+        end
+    end
+
+    @testset "zero-based Julian day" begin
+        # Note: Hard-coding years in tests below make failed testsets easier to read
+
+        # Non-leap year
+        # Note: For non-leap years day 365 is actually the first day of the next year which
+        # is why the test for `dayofyear` requires mod 365.
+        for d in [0; 58:60; 365]
+            f, i = _parsesub_tzdate("$d")
+            @test f(2019) == Date(2019) + Day(d)
+            @test dayofyear(f(2019)) == (d % 365) + 1
+        end
+
+        # Leap year
+        for d in [0; 58:60; 365]
+            f, i = _parsesub_tzdate("$d")
+            @test f(2020) == Date(2020) + Day(d)
+            @test dayofyear(f(2020)) == d + 1
+        end
+    end
+
+    @testset "month, week-of-month, day-of-week" begin
+        @testset "First Sunday in October" begin
+            f, i = _parsesub_tzdate("M10.1.0")
+            @test f(2020) == Date(2020, 10, 4)
+            @test f(2019) == Date(2019, 10, 6)
+            @test f(2018) == Date(2018, 10, 7)
+            @test f(2017) == Date(2017, 10, 1)
+            @test i == 8
+        end
+
+        @testset "Third Sunday in March" begin
+            f, i = _parsesub_tzdate("M3.3.0")
+            @test f(2020) == Date(2020, 3, 15)
+            @test f(2019) == Date(2019, 3, 17)
+            @test f(2018) == Date(2018, 3, 18)
+            @test f(2017) == Date(2017, 3, 19)
+            @test i == 7
+        end
+
+        @testset "Last Saturday in January" begin
+            # Note: 2017 - 2020 only have 4 Saturdays in January
+            f1, i = _parsesub_tzdate("M1.5.6")
+            @test i == 7
+
+            f2, i = _parsesub_tzdate("M1.4.6")
+            @test i == 7
+
+            @test f1(2020) == f1(2020) == Date(2020, 1, 25)
+            @test f1(2019) == f2(2019) == Date(2019, 1, 26)
+            @test f1(2018) == f2(2018) == Date(2018, 1, 27)
+            @test f1(2017) == f2(2017) == Date(2017, 1, 28)
+        end
+    end
+end
+
+@testset "_parsesub_tz" begin
+    dst_start_date_msg = "Unable to parse daylight saving start date. Expected date and instead found end of string"
+    dst_start_time_msg = "Expected daylight saving start time and instead found end of string"
+    dst_end_missing_msg = "Expected to find daylight saving end and instead found end of string"
+    dst_end_date_msg = "Unable to parse daylight saving end date. Expected date and instead found end of string"
+
+    @testset "empty" begin
+        @test _parsesub_tz("") == (FixedTimeZone("UTC"), 1)
+    end
+
+    @testset "standard time only" begin
+        @test _parsesub_tz("FOO") == ParseNextError("Expected standard offset and instead found end of string", "FOO", 4, 3)
```

Comment:
Does not need to be included in here, but what are your thoughts on storing error messages in `src/exceptions.jl`, that way we can hard code these strings in one area and just reference them throughout a package?

Reply:
The exceptions in the package itself aren't ever replicated so it would only be for the benefit of the tests. In some cases with the tests I reference the error messages via a variable which has a similar effect.

I think having constants in the source would probably make the code harder to read so I'm inclined to keep things as is.